src/utils.ts

```diff
@@ -124,24 +129,32 @@ export const recursivelyDownloadTx = async (
 
   if (parsedTx.parents.length > 2) {
     // We downloaded a block, we should ignore it
-    return recursivelyDownloadTx(blockId, txIds, data);
+    return recursivelyDownloadTx(blockId, blockHeight, txIds, data);
   }
 
-  // If the first_block from the downloaded tx is not from the block we are searching, we should ignore it.
-  if (meta.first_block !== blockId) {
-    return recursivelyDownloadTx(blockId, txIds, data);
+  // Transaction was already confirmed by a different block in the past
+  if (meta.first_block && meta.first_block !== blockId) {
+    const firstBlockResponse = await downloadTx(meta.first_block);
+
+    // If the transaction was confirmed by an older block, ignore it as it was
+    // already sent to the wallet-service
+    if (firstBlockResponse.tx.height < blockHeight) {
+      return recursivelyDownloadTx(blockId, blockHeight, txIds, data);
+    }
   }
 
-  // check if we have already downloaded the parents
-  const newParents = parsedTx.parents.filter(parent => {
+  const inputList = parsedTx.inputs.map((input) => input.txId);
+  const txList = [...parsedTx.parents, ...inputList];
+
+  // check if we have already downloaded any of the transactions of the new list
+  const newTxIds = txList.filter(transaction => {
```

Comment:
This algorithm is O(n^2). We can easily create a set in O(n) and then run this algorithm in O(n) since the search on the set will run in O(1). I think we should add a note explaining that it gets slower and slower as the DFS dives into the DAG. I also think we should create an issue to get this algorithm fixed soon.

Reply:
I think we can just refactor `txIds` as a Set so we don't have to create it on every iteration, I will do it in another PR

Added the comment you suggested on c290b6ee588542ecdf3c52ebe75c34c1099bd00b